src/rv32_template.c

```diff
@@ -4,42 +4,73 @@
  */
 
 /* Internal */
-RVOP(nop, { rv->X[rv_reg_zero] = 0; })
+RVOP(nop, { rv->X[rv_reg_zero] = 0; }, {RVCODEGEN(nop, {})})
 
 /* LUI is used to build 32-bit constants and uses the U-type format. LUI
  * places the U-immediate value in the top 20 bits of the destination
  * register rd, filling in the lowest 12 bits with zeros. The 32-bit
  * result is sign-extended to 64 bits.
  */
-RVOP(lui, { rv->X[ir->rd] = ir->imm; })
+RVOP(lui, { rv->X[ir->rd] = ir->imm; }, {RVCODEGEN(lui, {
+         emit_load_imm(state, RAX, ir->imm);
+         emit_store(state, S32, RAX, platform_parameter_registers[0],
+                    offsetof(struct riscv_internal, X) + 4 * ir->rd);
```

Comment:
I anticipate the presence of template-based code generation, which necessitates minimal descriptions of target instructions without the need for complete C source code.

Reply:
Does you means writing IR here, and this IR would be parsed and translatedin to the current C code by python script?

Reply:
Exactly. The template-based JIT code generation is mapped to the use of a custom domain-specific language (DSL), which is interpreted by our Python script. Having minimal descriptions of target instructions without the need for complete C source code makes sense for both maintainability and future machine code lifting. See [Rellume — Lift machine code to LLVM IR](https://github.com/aengelke/rellume),

Reply:
Meanwhile, you can include a brief documentation section at the beginning of `rv32_template.c` explaining the use of DSL during JIT code generation.